astropy/io/fits/tests/test_fsspec.py

```diff
@@ -0,0 +1,128 @@
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+"""Can `astropy.io.fits.open` access (remote) data using the fsspec package?
+"""
+from astropy.io import fits
+from astropy.nddata import Cutout2D
+from astropy.utils.compat.optional_deps import HAS_FSSPEC, HAS_S3FS  # noqa
+from astropy.utils.data import get_pkg_data_filename, conf
+
+import numpy as np
+from numpy.testing import assert_array_equal, assert_allclose
+import pytest
+
+
+@pytest.mark.skipif("not HAS_FSSPEC")
+def test_fsspec_local():
+    """Can we use fsspec to read a local file?"""
+    fn = get_pkg_data_filename('data/test0.fits')
+    hdulist_classic = fits.open(fn)
+    hdulist_fsspec = fits.open(fn, use_fsspec=True)
+    assert_array_equal(hdulist_classic[2].data, hdulist_fsspec[2].data)
+    assert_array_equal(hdulist_classic[2].section[3:5], hdulist_fsspec[2].section[3:5])
+    hdulist_classic.close()
+    hdulist_fsspec.close()
+
+
+@pytest.mark.skipif("not HAS_FSSPEC")
+def test_fsspec_local_write(tmpdir):
+    """Can we write to a local file that was opened using fsspec?"""
+    fn = get_pkg_data_filename('data/test0.fits')
+    fn_tmp = tmpdir / "tmp.fits"
+    with fits.open(fn, use_fsspec=True) as hdul:
+        # writing to a section is never allowed
+        with pytest.raises(TypeError):
+            hdul[1].section[0, 0] = -999
+        # however writing to .data should work
+        hdul[1].data[2, 3] = -999
+        assert hdul[1].data[2,3] == -999
+        hdul.writeto(fn_tmp)
+
+    # Is the new value present when we re-open the file?
+    with fits.open(fn_tmp) as hdul:
+        assert hdul[1].data[2,3] == -999
+
+    # Does fsspec support `mode="update"` for local files?
+    with fits.open(str(fn_tmp), use_fsspec=True, mode="update") as hdul:
+        hdul[1].data[2,3] = 42
+    with fits.open(fn_tmp) as hdul:
+        assert hdul[1].data[2,3] == 42
+
+
+@pytest.mark.remote_data
+@pytest.mark.skipif("not HAS_FSSPEC")
+def test_fsspec_http():
+    """Can we use fsspec to open a remote FITS file via http?"""
+    uri = "https://mast.stsci.edu/api/v0.1/Download/file/?uri=mast:HST/product/j8pu0y010_drc.fits"
+    # Expected array was obtained by downloading the file locally and executing:
+    # with fits.open(local_path) as hdul:
+    #     hdul[1].data[1000:1002, 2000:2003]
+    expected = np.array([[0.00600598, 0.00546154, 0.00134362],
+                         [0.00073434, 0.00825543, 0.0059365]])
+    with fits.open(uri, use_fsspec=True) as hdul:
+        # Do we retrieve the expected array?
+        assert_allclose(hdul[1].section[1000:1002, 2000:2003], expected, atol=1e-7)
+        # The file has multiple extensions which are not yet downloaded;
+        # the repr and string representation should reflect this.
+        assert "partially read" in repr(hdul)
+        assert "partially read" in str(hdul)
+
+    # Can the user also pass an fsspec file object directly to fits open?
+    import fsspec
```

Comment:
maybe move the import from here and below to the top under an  `if HAS_FSSPEC` conditional.

Reply:
Note: moving the import to the top under `HAS_FSSPEC` would cause `fsspec` to be imported whenever it is installed, even when it is not *used*.

Leaving the import where it is now means `fsspec` will only be imported when it is actually *used*.

@bsipocz Do you prefer moving it to the top regardless?  I'm happy to do either because fsspec is a cheap import.

Reply:
Previous discussion on this very topic: https://github.com/astropy/astropy/pull/13238#discussion_r869296062

Reply:
I fully agree with the previous comment, and only suggest doing this move for the test file where we end up using it every time it's available.

Reply:
Oops I didn't notice it is the test file you commented on.  Apologies!!

Yes, :100:% agreed, better to move it to the top in the test file!

Reply:
Fixed via 7e7d668852e800d5fce640c020309ccbc7ea4607.  Thank you!